File: legacy/isma-core/src/main/java/ru/nstu/isma/core/sim/fdm/FDMIndexIterator.java
package ru.nstu.isma.core.sim.fdm;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * by
 * Bessonov Alex.
 * Date: 13.12.13 Time: 1:43
 * Итератор перебора всех вариантов индекса
 */

public class FDMIndexIterator {
    private ArrayList<FDMIndexedApxVar> indexes = new ArrayList<FDMIndexedApxVar>();

    private Map<String, FDMIndexedApxVar> nameMapping = new HashMap<String, FDMIndexedApxVar>();

    boolean readyToFight = false;

    public void addIndex(FDMIndexedApxVar v) {
        indexes.add(v);
        nameMapping.put(v.getCode(), v);
        readyToFight = false;
    }

    public List<FDMIndexedApxVar> start() {
        for (FDMIndexedApxVar v : indexes) {
            v.setIndex(1);
        }
        readyToFight = true;
        return indexes;
    }

    public List<FDMIndexedApxVar> next() {
        if (!readyToFight) {
            throw new RuntimeException("Iterator not ready! Call 'start()' method before!");
        }
        // предотвращаем выход за границу
        if (atEnd()) {
            throw new RuntimeException("Iterator at the end!");
        }
        int curIdx = 0;
        boolean success = false;
        FDMIndexedApxVar cur;
        while(!success) {
            cur = indexes.get(curIdx);
            cur.inc();
            if (!cur.isValid()) {
                cur.setIndex(1);
                curIdx++;
            } else {
                success = true;
            }
        }
        return indexes;
    }

    public ArrayList<FDMIndexedApxVar> get() {
        return indexes;
    }

    public FDMIndexedApxVar getIndex(String code) {
        return nameMapping.get(code);
    }

    public boolean atEnd() {
        for (FDMIndexedApxVar v : indexes) {
            if (!v.isMax()) {
                return false;
            }
        }
        return true;
    }

}

File: isma-next-core/src/main/kotlin/ru/nstu/isma/next/core/sim/fdm/FDMIndexIterator.kt
package ru.nstu.isma.next.core.sim.fdm

import java.util.*

/**
 * by
 * Bessonov Alex.
 * Date: 13.12.13 Time: 1:43
 * Итератор перебора всех вариантов индекса
 */
class FDMIndexIterator {
    private val indexes = ArrayList<FDMIndexedApxVar>()
    private val nameMapping: MutableMap<String, FDMIndexedApxVar> = HashMap()
    var readyToFight = false
    fun addIndex(v: FDMIndexedApxVar) {
        indexes.add(v)
        nameMapping[v.code] = v
        readyToFight = false
    }

    fun start(): List<FDMIndexedApxVar> {
        for (v in indexes) {
            v.setIndex(1)
        }
        readyToFight = true
        return indexes
    }

    operator fun next(): List<FDMIndexedApxVar> {
        if (!readyToFight) {
            throw RuntimeException("Iterator not ready! Call 'start()' method before!")
        }
        // предотвращаем выход за границу
        if (atEnd()) {
            throw RuntimeException("Iterator at the end!")
        }
        var curIdx = 0
        var success = false
        var cur: FDMIndexedApxVar
        while (!success) {
            cur = indexes[curIdx]
            cur.inc()
            if (!cur.isValid) {
                cur.setIndex(1)
                curIdx++
            } else {
                success = true
            }
        }
        return indexes
    }

    fun get(): ArrayList<FDMIndexedApxVar> {
        return indexes
    }

    fun getIndex(code: String): FDMIndexedApxVar? {
        return nameMapping[code]
    }

    fun atEnd(): Boolean {
        for (v in indexes) {
            if (!v.isMax) {
                return false
            }
        }
        return true
    }
}